numba/tests/test_dictobject.py

```diff
@@ -1186,6 +1186,134 @@ def test_str(self):
+            return Dict(items)
+
+        @njit
+        def func7(k, v):
+            return Dict({k: v})
+
+        vals = (
+            (func1, [(0, 1, 2), 'abc'], Dict({0: 'a', 1: 'b', 2: 'c'})),
+            (func2, [(0, 1, 2), 'abc'], Dict({0: 'a', 1: 'b', 2: 'c'})),
+            (func3, [(0, 1, 2), 'abc'], Dict({0: 'a', 1: 'b', 2: 'c'})),
+            (func4, [(0, 1, 2), 'abc'], Dict(
+                {0: ((0, 'a'), 0), 1: ((1, 'b'), 1), 2: ((2, 'c'), 2)})),
+            (func5, [(0, 1, 2), 'abc'], Dict(
+                {(0, 'a'): 'a', (1, 'b'): 'b', (2, 'c'): 'c'})),
+            (func6, [(),], Dict({})),
+            (func6, [((1, 'a'), (3, 'b')),], Dict({1: 'a', 3: 'b'})),
+            (func7, ['key', Dict({1: 'abc'})], Dict({'key': Dict({1: 'abc'})})),
+            (func7, ['key', List([1, 2, 3])], Dict({'key': List([1, 2, 3])})),
+        )
+
+        for func, args, expected in vals:
+            if self.jit_enabled:
+                got = func(*args)
+            else:
+                got = func.py_func(*args)
+            self.assertPreciseEqual(expected, got)
+
+
+class TestDictIterableCtorJit(DictIterableCtor):
+
+    def setup(self):
+        self.jit_enabled = True
+
+    def test_exception_no_iterable_arg(self):
+        @njit
+        def ctor():
+            return Dict(3)
+
+        msg = ".*No implementation of function.*"
+        with self.assertRaisesRegex(TypingError, msg):
+            ctor()
+
+    def test_exception_mapping_arg(self):
+        @njit
+        def ctor():
+            return Dict({1: 2})
+
+        msg = ".*No implementation of function.*"
+        with self.assertRaisesRegex(TypingError, msg):
+            ctor()
+
+    def test_exception_setitem(self):
+        @njit
+        def ctor():
+            return Dict(((1, 'a'), (2, 'b', 3)))
+
+        msg = ".*No implementation of function.*"
+        with self.assertRaisesRegex(TypingError, msg):
+            ctor()
+
+
+class TestDictIterableCtorNoJit(DictIterableCtor):
+
+    def setup(self):
+        self.jit_enabled = False
+
+    def test_exception_nargs(self):
+        msg = 'Dict expect at most 1 argument, got 2'
+        with self.assertRaisesRegex(TypeError, msg):
+            Dict(1, 2)
+
+    def test_exception_non_iterable_arg(self):
+        msg = '.*object is not iterable.*'
+        with self.assertRaisesRegex(TypeError, msg):
+            Dict(3)
+
+    def test_exception_setitem(self):
+        msg = ".*dictionary update sequence element #1 has length 3.*"
+        with self.assertRaisesRegex(ValueError, msg):
+            Dict(((1, 'a'), (2, 'b', 3)))
+
+
+class TestDictMappingCtor(TestCase):
+
+    def test_basic(self):
+        d = Dict({1: 2, 3: 4, 5: 6})
+        self.assertEqual(len(d), 3)
+        self.assertEqual(len(d.values()), 3)
+
+    def test_exception(self):
+        msg = '.*No implementation of function.*'
+        with self.assertRaisesRegex(TypingError, msg):
+            Dict({1: 2, 'a': 'b'})
+
+
```

Comment:
Here's some things that either aren't working or could work better:
```python
from numba import njit
import numpy as np

# Valid, won't compile, fails at typing.
@njit
def foo():
    return dict(([1, 2], [1, 2]))

print(foo())


#Wrong error message, the issue is the types in the iterable aren't supported
#whereas the error message complains of: `Unsupported use of `dict()` with
#positional or keyword argument(s). The only supported uses are `dict()` or
#`dict(*iterable)`.`
@njit
def foo():
    return dict(((1, 2), (3, 'a')))

print(foo())


#Lowering error, should be caught earlier
@njit
def foo():
    return dict((('a', 'b', 'c'), ('d', 'e', 'f')))

print(foo())


#This should really complain about unhashable key type
@njit
def foo():
    return dict((('a', 1), ({}, 2)))

print(foo())


# Mapping support not working under JIT
@njit
def foo():
    return dict({x: x ** 2 for x in range(3)})

print(foo())
```

Reply:
I believe I have address all the issues pointed out in our comment, except for the last one, which involves `dict(mapping)`.